src/snakemake/workflow.py

```
@@ -2004,6 +2030,42 @@ def check_may_use_software_deployment(method):
+                if not (
+                    ruleinfo.script
+                    or ruleinfo.wrapper
+                    or ruleinfo.shellcmd
+                    or ruleinfo.notebook
+                ):
+                    raise RuleException(
+                        "Software directive is only allowed "
+                        "with shell, script, notebook or wrapper directives "
+                        "(not with run or template_engine).",
+                        rule=rule,
+                    )
```

Ruff (0.14.6)

Avoid specifying long messages outside the exception class

(TRY003)